File: drills/elixir/solutions/nc07_longest_consecutive__sorting.exs
defmodule Solution do
  def longest_consecutive(nums) do
    # No set: sort, then fold once over the gaps. O(n log n) rather than O(n),
    # but the run logic reads straight through.
    case Enum.sort(nums) do
      [] ->
        0

      [first | rest] ->
        {_previous, _run, best} =
          Enum.reduce(rest, {first, 1, 1}, fn n, {previous, run, best} ->
            case n - previous do
              0 -> {previous, run, best}
              1 -> {n, run + 1, max(best, run + 1)}
              _ -> {n, 1, best}
            end
          end)

        best
    end
  end
end

File: drills/go/solutions/nc07_longest_consecutive__sorting.go
package main

import "sort"

func longestConsecutive(nums []int) int {
	if len(nums) == 0 {
		return 0
	}
	sorted := append([]int(nil), nums...)
	sort.Ints(sorted)
	longest, length := 1, 1
	for i := 1; i < len(sorted); i++ {
		switch {
		case sorted[i] == sorted[i-1]:
			continue
		case sorted[i] == sorted[i-1]+1:
			length++
		default:
			length = 1
		}
		if length > longest {
			longest = length
		}
	}
	return longest
}
